debian/rednotebook.1
.\" Hey, EMACS: -*- nroff -*-
.TH "REDNOTEBOOK" "1" "2013-10-18" "Jonathan Wiltshire" ""
.SH "NAME"
rednotebook \- daily journal with calendar, templates and keyword searching
.SH "SYNOPSIS"
.B rednotebook [\fIoptions\fP] [\fIjournal-path\fP]
.SH "DESCRIPTION"
.B RedNotebook
is a desktop diary that makes it very easy for you to keep track of the stuff you do and the thoughts you have. This journal software helps you to write whole passages or just facts and does so in style.
.SH "OPTIONS"
.LP
.TP
\fB\-\-date=\fIDATE\fR
Open RedNotebook on the specified date (format YYYY-MM-DD)
.TP
\fB\-h, \-\-help\fR
Show help text and exit
.TP
\fB\-\-version
Show version number and exit
.LP
\fBjournal-path\fR can be one of the following:
 \- An absolute path (e.g. /home/username/myjournal)
 \- A relative patch (e.g. ../dir/myjournal)
 \- The name of a directory under $HOME/.rednotebook/ (e.g. myjournal)
.LP
If the \fBjournal-path\fR is omitted the last session's journal will be used.
At the first program start this defaults to \fI$HOME/.rednotebook/data\fR.
.LP
Configuration is stored in the file \fI~/.rednotebook/configuration.cfg\fR.
.SH "FILES"
\fI/usr/bin/rednotebook\fR \- \fBRedNotebook\fR executable
.br
\fI~/.rednotebook\fR \- per\-user configuration and notebooks
.SH "AUTHOR"
rednotebook was written by Jendrik Seipp <jendrikseipp@gmail.com>.
.PP
This manual page was written by Jonathan Wiltshire <debian@jwiltshire.org.uk>,
for the Debian project (and may be used by others).


CHANGELOG
== 2.2 (2017-09-08) ==
* Port RedNotebook 2 to Windows.
* Windows: uninstall old version before installing new version to remove old files.
* Windows: use Aspell for spell checking.
* Update Debian files (@kathenas).

== 2.1.5 (2017-08-09) ==
* Fix debian/control.

== 2.1.4 (2017-08-08) ==
* Use old names for appdata and desktop files.

== 2.1.3 (2017-08-06) ==
* Fix creating translation files.

== 2.1.2 (2017-08-03) ==
* Fix "Exec" field in .desktop file.

== 2.1.1 (2017-08-02) ==
* Reset package name to 'rednotebook'.

== 2.1 (2017-07-29) ==
* Make webkit optional (but highly recommended). If missing, show preview in external browser.
* Support inserting SVG images.
* Don't switch between edit and review mode automatically by default.

== 2.0 (2017-05-19) ==
* Port to Python 3 and GTK 3.
* Add index of tags to LaTeX export (#324, thanks Alex Schickedanz).
* Use new CDN link for MathJax.

== 1.15 (2017-02-11) ==
* Bundle pygtkspellcheck 4.0.5 since earlier versions contain a bug (lp:1615629).
* Fix toggling autostart (lp:1628497).
* Set system tray icon name (lp:1660129).

== 1.14 (2016-09-26) ==
* Use new pygtkspellcheck API (lp:1592727).
* Fix conversion from old single "Tags" category to new tags format.

== 1.13 (2016-06-17) ==
* When selecting a journal directory, show all journal directories by default (thanks Paul Jackson).
* Allow ampersands in e-mail addresses (lp:1570476, thanks pdofak).
* python-gtkspell has been renamed to python-gtkspellcheck. Update docs and debian/control accordingly.

== 1.12 (2016-03-28) ==
* Add option to hide right-hand tags pane (thanks Ron Brown, Jr.).
* Never overwrite externally changed month files (thanks Felix Zörgiebel).
* Remove support for Python 2.6.

== 1.11 (2015-11-08) ==
* Remind users to make new backup if last backup is older than a month.
* When inserting a link for a selected text passage, replace text passage.

== 1.10.4 (2015-10-19) ==
* Fix right-clicking cloud words to hide them.

== 1.10.3 (2015-10-15) ==
* Make #hashtags regular expression faster (up to 1000x).

== 1.10.2 (2015-08-24) ==
* Fix Chinese fonts in preview mode on Windows by using the fonts mingliu and MS Mincho (thanks Amos Ng).
* Fix instructions for running RedNotebook on Windows (thanks Amos Ng).
* Move repository from launchpad to github.

== 1.10.1 (2015-04-14) ==
* When undoing a formatting action, only remove formatting, not the text (lp:1326606).
* Fix searching with enter for text with non-ASCII characters (lp:1430697).
* Fix filtering exports by tags containing non-ASCII characters (lp:1267263).
* Don't remove whitespace from old-style tags in autocomplete box (lp:1414603).
* Don't try to change spellchecking language if spellchecker is not set up (lp:1443818).
* Don't try to access files with wrong encoding (lp:1443818).

== 1.10 (2015-04-12) ==
* Write data to temporary files first to prevent corrupted month files.
* Add format button for monospace font. Automatically add the correct format for code blocks.
* Show warning for dates before 1900 (not supported by Python's datetime module).
* Fix undo/redo.
* Don't add unneeded newlines around titles and code.

== 1.9.0 (2014-12-27) ==
* Add #tags to cloud ignore list to remove them from the tag cloud (thanks Przemysław Buczkowski).
* Remove option to start RedNotebook minimized.
* Do not remove menu bar in fullscreen mode (lp:1400356).
* Allow exiting fullscreen mode with ESC key.
* Only allow comment signs (#) at the beginning of a line in the config file.
* Fix reading configuration files.

== 1.8.1 (2014-08-03) ==
* Do not show new version dialog if latest version cannot be determined (lp:1324126).
* Fix date formatting for invalid locale encodings.
* Add necessary library files for spell-checking on Windows (lp:1331876).

== 1.8.0 (2013-12-12) ==
* Add font selection for edit mode (Philip Akesson).
* Allow changing preview and cloud font in preferences.
* Only allow opening RedNotebook minimized on Windows since other systems may lack a system tray.
* Fix: Display tags starting with "SEP" in preview (lp:1255582).
* Write scripts to cross-compile RedNotebook Windows exe and installer on Linux.

== 1.7.3 (2013-11-10) ==
* Jump to a specific date on startup with "--date 2013-10-31" on the commandline (Rob Norris).
* Display current date in the title bar (Rob Norris).
* Support inserting multiple pictures in one step.
* Add DejaVu Sans as font fallback for clouds and preview.
* Fix setting last image directory for insert dialog.
* Only allow exporting selected text if we are in edit mode (lp:1221792).

== 1.7.2 (2013-06-28) ==
* Show error if saving fails due to a directory not being created.
* Ignore hashtags starting with more than one #.
* Fix: Allow inserting files and pictures from the "recently used" section (lp:1195759).

== 1.7.1 (2013-03-01) ==
* Fix: Insert spellchecking correction in the correct position (LP:1137925).

== 1.7.0 (2013-02-28) ==
* Allow filtering exported days by tags (Alistair Marshall).
* Add option to export only the currently selected text (Alistair Marshall).
* Move spellcheck option from preferences to edit menu and add F7 shortcut (Alistair Marshall).
* Enable spellchecking on Windows. See help for adding custom dictionaries.
* Better error message for invalid markup.
* Show warning if no directory is selected before clicking the "Open" button when choosing a journal directory.
* Gracefully handle BadStatusLines when checking for new versions.

rednotebook/redNotebook.py
# -*- coding: utf-8 -*-
# -----------------------------------------------------------------------
# Copyright (c) 2009  Jendrik Seipp
# 
# RedNotebook is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# 
# RedNotebook is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License along
# with RedNotebook; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
# -----------------------------------------------------------------------

from __future__ import with_statement

import sys
import datetime
import os
import operator
import collections
import time
from optparse import OptionParser, OptionValueError


if hasattr(sys, "frozen"):
	from rednotebook.util import filesystem
	from rednotebook.util import utils
	from rednotebook import info
	from rednotebook import configuration
else:
	from util import filesystem # creates a copy of filesystem module
	#import util.filesystem # imports the original filesystem module
	from util import utils
	import info
	import configuration
	
	

def parse_options():
	parser = OptionParser(usage="usage: %prog [options] [journal-path]",
						  description=info.command_line_help,
						  #option_class=ExtOption,
						  formatter=utils.IndentedHelpFormatterWithNL(),
						  )
	
	parser.add_option(
		'-d', '--debug', dest='debug', \
		default=False, action='store_true',
		help='Output debugging messages ' \
			'(default: False)')
	
	parser.add_option(
		'-m', '--minimized', dest='minimized', \
		default=False, action='store_true',
		help='Start mimimized to system tray ' \
			'(default: False)')
	
	options, args = parser.parse_args()
		
	return options, args

options, args = parse_options()



## ---------------------- Enable logging -------------------------------

import logging

def setup_logging(log_file):
	logging_levels = {'debug': logging.DEBUG,
					'info': logging.INFO,
					'warning': logging.WARNING,
					'error': logging.ERROR,
					'critical': logging.CRITICAL}
	
	# File logging
	if sys.platform == 'win32' and hasattr(sys, "frozen"):
		utils.redirect_output_to_file(log_file)
	
	file_logging_stream = open(log_file, 'w')
	
	# We want to have the error messages in the logfile
	sys.stderr = utils.StreamDuplicator(sys.__stderr__, [file_logging_stream])
	
	# Write a log containing every output to a log file
	logging.basicConfig(level=logging.DEBUG,
						format='%(asctime)s %(levelname)-8s %(message)s',
						#filename=filesystem.log_file,
						#filemode='w',
						stream=file_logging_stream,#sys.stdout,
						)
	
	level = logging.INFO
	#if len(sys.argv) > 1:
		#level = logging_levels.get(sys.argv[1], level)
	if options.debug:
		level = logging.DEBUG
	
	# define a Handler which writes INFO messages or higher to the sys.stdout
	console = logging.StreamHandler(sys.stdout)
	console.setLevel(level)
	# set a format which is simpler for console use
	formatter = logging.Formatter('%(levelname)-8s %(message)s')
	# tell the handler to use this format
	console.setFormatter(formatter)
	# add the handler to the root logger
	logging.getLogger('').addHandler(console)
	
	logging.debug('sys.stdout logging level: %s' % level)
	logging.info('Writing log to file "%s"' % log_file)


default_config_file = os.path.join(filesystem.app_dir, 'files', 'default.cfg')
default_config = configuration.Config(default_config_file)

dirs = filesystem.Filenames(default_config)
setup_logging(dirs.log_file)

## ------------------ end Enable logging -------------------------------



## ---------------------- Enable i18n -------------------------------

# set the locale for all categories to the user’s default setting 
# (typically specified in the LANG environment variable)
import locale
lang = os.environ.get('LANG', None)
logging.debug('LANG: %s' % lang)
default_locale = locale.getdefaultlocale()[0]
logging.debug('Default locale: %s' % default_locale)
try:
	locale.setlocale(locale.LC_ALL, '')
	logging.debug('Set default locale: "%s"' % default_locale)
except locale.Error, err:
	# unsupported locale setting
	logging.error('Locale "%s" could not be set: "%s"' % (default_locale, err))
	logging.error('Probably you have to install the appropriate language packs')

# If the default locale could be determined and the LANG env variable
# has not been set externally, set LANG to the default locale
# This is necessary only for windows where program strings are not
# shown in the system language, but in English
if default_locale and not lang:
	logging.debug('Setting LANG to %s' % default_locale)
	os.environ['LANG'] = default_locale
	
LOCALE_PATH = os.path.join(dirs.app_dir, 'i18n')

# the name of the gettext domain. because we have our translation files
# not in a global folder this doesn't really matter, setting it to the
# application name is a good idea tough.
GETTEXT_DOMAIN = 'rednotebook'

# set up the gettext system
import gettext

# Adding locale to the list of modules translates gtkbuilder strings
modules = [gettext, locale]

# Sometimes this doesn't work though, 
# so we try to call gtk.glade's function as well if glade is present
try:
	import gtk.glade
	modules.append(gtk.glade)
except ImportError, err:
	pass

for module in modules:
	try:
		# locale.bintextdomain and locale textdomain not available on win
		module.bindtextdomain(GETTEXT_DOMAIN, LOCALE_PATH)
		module.textdomain(GETTEXT_DOMAIN)
	except AttributeError, err:
		logging.info(err)
		
# register the gettext function for the whole interpreter as "_"
import __builtin__
__builtin__._ = gettext.gettext



## ------------------- end Enable i18n -------------------------------




try:
	import pygtk
	if not sys.platform == 'win32':
		pygtk.require("2.0")
except ImportError:
	logging.error('pygtk not found. Please install PyGTK (python-gtk2)')
	sys.exit(1)

try:
	import gtk
	
	import gobject
	# Some notes on threads_init:
	# only gtk.gdk.threads_init(): pdf export works, but gui hangs afterwards
	# only gobject.threads_init(): pdf export works, gui works
	# both: pdf export works, gui hangs afterwards	
	gobject.threads_init() # only initializes threading in the glib/gobject module
	#gtk.gdk.threads_init() # also initializes the gdk threads
except (ImportError, AssertionError), e:
	logging.error(e)
	logging.error('gtk not found. Please install PyGTK (python-gtk2)')
	sys.exit(1)

	

logging.info('AppDir: %s' % filesystem.app_dir)
base_dir = os.path.abspath(os.path.join(filesystem.app_dir, '../'))
logging.info('BaseDir: %s' % base_dir)
if base_dir not in sys.path:
	# Adding BaseDir to sys.path
	sys.path.insert(0, base_dir)
	

# This version of import is needed for win32 to work
from rednotebook.util import unicode
from rednotebook.util import dates
#from rednotebook import info
#from rednotebook import configuration
from rednotebook import backup


from rednotebook.util.statistics import Statistics
from rednotebook.gui.mainWindow import MainWindow
from rednotebook.storage import Storage
#import rednotebook.storage
from rednotebook.data import Month

class Journal:
	
	def __init__(self):
		self.dirs = dirs
		
		user_config = configuration.Config(self.dirs.config_file)
		# Apply defaults where no custom values have been set
		for key, value in default_config.items():
			if key not in user_config:
				user_config[key] = value
		self.config = user_config
		self.config.save_state()
		
		logging.info('Running in portable mode: %s' % self.dirs.portable)
		
		self.testing = False
		if options.debug:
			self.testing = True
			logging.debug('Debug Mode is on')
			
		# Allow starting minimized to tray
		# When we start minimized we have to set the tray icon visible
		self.start_minimized = options.minimized
		if self.start_minimized:
			self.config['closeToTray'] = 1
		
		self.month = None
		self.date = None
		self.months = {}
		
		# The dir name is the title
		self.title = ''
		
		# show instructions at first start
		logging.info('First Start: %s' % self.dirs.is_first_start)
		
		logging.info('RedNotebook version: %s' % info.version)
		logging.info(filesystem.get_platform_info())
		
		utils.set_environment_variables(self.config)
		
		self.actual_date = datetime.date.today()
		
		# Let components check if the MainWindow has been created
		self.frame = None
		self.frame = MainWindow(self)
		
		self.storage = Storage()
		self.open_journal(self.get_journal_path())
		
		self.archiver = backup.Archiver(self)
		
		# Check for a new version
		if self.config.read('checkForNewVersion', default=0) == 1:
			utils.check_new_version(self.frame, info.version, startup=True)
			
		# Automatically save the content after a period of time
		if not self.testing:
			gobject.timeout_add_seconds(600, self.save_to_disk)
			
	
	def get_journal_path(self):
		'''
		Retrieve the path from optional args or return standard value if args
		not present
		'''
		if not args:
			data_dir = self.config.read('dataDir', self.dirs.data_dir)
			if not os.path.isabs(data_dir):
				data_dir = os.path.join(self.dirs.app_dir, data_dir)
				data_dir = os.path.normpath(data_dir)
			return data_dir
		
		# path_arg can be e.g. data (under .rednotebook), data (elsewhere), 
		# or an absolute path /home/username/myjournal
		# Try to find the journal under the standard location or at the given
		# absolute or relative location
		path_arg = args[0]
		
		logging.debug('Trying to find journal "%s"' % path_arg)
		
		paths_to_check = [path_arg, os.path.join(self.dirs.rednotebook_user_dir, path_arg)]
		
		for path in paths_to_check:
			if os.path.exists(path):
				if os.path.isdir(path):
					return path
				else:
					logging.warning('To open a journal you must specify a '
								'directory, not a file.')
		
		logging.error('The path "%s" is no valid journal directory. ' 
					'Execute "rednotebook -h" for instructions' % path_arg)
		sys.exit(1)
			
			
	
	
	def get_days_in_date_range(self, range):
		start_date, end_date = range
		assert start_date <= end_date
		
		sorted_days = self.sorted_days
		days_in_date_range = []
		for day in sorted_days:
			if day.date < start_date:
				continue
			elif day.date >= start_date and day.date <= end_date:
				days_in_date_range.append(day)
			elif day.date > end_date:
				break
		return days_in_date_range
		
		
	def _get_sorted_days(self):
		return sorted(self.days, key=lambda day: day.date)
	sorted_days = property(_get_sorted_days)
	
	
	def get_edit_date_of_entry_number(self, entry_number):
		sorted_days = self.sorted_days
		if len(self.sorted_days) == 0:
			return datetime.date.today()
		return self.sorted_days[entry_number % len(sorted_days)].date
	
	
	def backup_contents(self, backup_file):
		self.save_to_disk()
		
		if backup_file:
			self.archiver.backup(backup_file)
			
		
	def exit(self):
		self.frame.add_values_to_config()
		
		# Make it possible to stop the program from exiting
		# e.g. if the journal could not be saved
		self.is_allowed_to_exit = True
		self.save_to_disk(exit_imminent=True)
		
		if self.is_allowed_to_exit:
			logging.info('Goodbye!')
			gtk.main_quit()

	
	def save_to_disk(self, exit_imminent=False, changing_journal=False, saveas=False):
		logging.info('Trying to save the journal')
		
		self.save_old_day()
			
		try:
			filesystem.make_directory(self.dirs.data_dir)
		except OSError, err:
			self.frame.show_save_error_dialog(exit_imminent)
			return True
			
		if not os.path.exists(self.dirs.data_dir):
			logging.error('Save path does not exist')
			self.frame.show_save_error_dialog(exit_imminent)
			return True
			
		
		something_saved = self.storage.save_months_to_disk(self.months, \
			self.dirs.data_dir, self.frame, exit_imminent, changing_journal, saveas)					
		
		if something_saved:
			self.show_message(_('The content has been saved to %s') % self.dirs.data_dir, error=False)
		else:
			self.show_message(_('Nothing to save'), error=False)
		
		if self.config.changed():
			try:
				filesystem.make_directory(self.dirs.red_notebook_user_dir)
				self.config.save_to_disk()
			except IOError, err:
				self.show_message(_('Configuration could not be saved. Please check your permissions'))
		
		if not (exit_imminent or changing_journal) and something_saved:
			# Update cloud
			self.frame.cloud.update(force_update=True)
			
		# tell gobject to keep saving the content in regular intervals
		return True
	
	
	def open_journal(self, data_dir, load_files=True):
		
		if self.months:
			self.save_to_disk(changing_journal=True)
			
		# Password Protection
		#password = self.config.read('password', '')
		
		logging.info('Opening journal at %s' % data_dir)
		
		if not os.path.exists(data_dir):
			logging.warning('The data dir %s does not exist. Select a different dir.' \
						% data_dir)
			
			self.frame.show_dir_chooser('open', dir_not_found=True)
			return
		
		data_dir_empty = not os.listdir(data_dir)
		
		if not load_files and not data_dir_empty:
			msg_part1 = _('The selected folder is not empty.')
			msg_part2 = _('To prevent you from overwriting data, the folder content has been imported into the new journal.')
			self.show_message('%s %s' % (msg_part1, msg_part2), error=False)
		elif load_files and data_dir_empty:
			self.show_message(_('The selected folder is empty. A new journal has been created.'), \
								error=False)
		
		self.dirs.data_dir = data_dir
		
		self.month = None
		self.months.clear()
		
		# We always want to load all files
		if load_files or True:
			self.months = self.storage.load_all_months_from_disk(data_dir)
		
		# Nothing to save before first day change
		self.load_day(self.actual_date)
		
		self.stats = Statistics(self)
		
		sorted_categories = sorted(self.node_names, key=lambda category: str(category).lower())
		self.frame.categories_tree_view.categories = sorted_categories
		
		if self.dirs.is_first_start and data_dir_empty:
			logging.info('Adding example content')
			self.add_instruction_content()
			
		# Notebook is only on page 1 here, if we are opening a journal the second time
		if self.frame.search_notebook.get_current_page() == 1:
			# We have opened a new journal
			self.frame.cloud.update(force_update=True)
		else:
			# Show cloud tab, cloud is updated automatically
			self.frame.search_notebook.set_current_page(1)
		
		# Reset Search
		self.frame.search_box.clear()
		
		self.title = filesystem.get_journal_title(data_dir)
		
		# Set frame title
		if self.title == 'data':
			frame_title = 'RedNotebook'
		else:
			frame_title = 'RedNotebook - ' + self.title
		self.frame.main_frame.set_title(frame_title)
		
		# Save the folder for next start
		if not self.dirs.portable:
			self.config['dataDir'] = data_dir
		else:
			rel_data_dir = filesystem.get_relative_path(self.dirs.appDir, data_dir)
			self.config['dataDir'] = rel_data_dir
	
		
	def get_month(self, date):
		'''
		Returns the corresponding month if it has previously been visited,
		otherwise a new month is created and returned
		'''
		
		year_and_month = dates.get_year_and_month_from_date(date)
		
		# Selected month has not been loaded or created yet
		if not self.months.has_key(year_and_month):
			self.months[year_and_month] = Month(date.year, date.month)
			
		return self.months[year_and_month]
	
	
	def save_old_day(self):
		'''Order is important'''
		old_content = self.day.content
		self.day.content = self.frame.categories_tree_view.get_day_content()
		self.day.text = self.frame.get_day_text()
		
		content_changed = not (old_content == self.day.content)
		if content_changed:
			self.month.edited = True
		
		self.frame.calendar.set_day_edited(self.date.day, not self.day.empty)
	
	
	def load_day(self, new_date):
		old_date = self.date
		self.date = new_date
		
		if not Month.same_month(new_date, old_date) or self.month is None:
			self.month = self.get_month(self.date)
			#self.month.visited = True
		
		self.frame.set_date(self.month, self.date, self.day)
		
		
	def merge_days(self, days):
		'''
		Method used by importers
		'''
		self.save_old_day()
		for new_day in days:
			date = new_day.date
			month = self.get_month(date)
			old_day = month.get_day(date.day)
			old_day.merge(new_day)
			month.edited = True
			
		
	def _get_current_day(self):
		return self.month.get_day(self.date.day)
	day = property(_get_current_day)
	
	
	def change_date(self, new_date):
		if new_date == self.date:
			return
		
		self.save_old_day()
		self.load_day(new_date)
		
		
	def go_to_next_day(self):
		self.change_date(self.date + dates.one_day)
		
		
	def go_to_prev_day(self):
		self.change_date(self.date - dates.one_day)
			
			
	def show_message(self, message_text, error=False, countdown=True):
		self.frame.statusbar.show_text(message_text, error, countdown)
		logging.info(message_text)
		
		
	def _get_node_names(self):
		node_names = set([])
		for month in self.months.values():
			node_names |= set(month.node_names)
		return list(node_names)
	node_names = property(_get_node_names)
	
	
	def _get_tags(self):
		tags = set([])
		for month in self.months.values():
			tags |= set(month.tags)
		return list(tags)
	tags = property(_get_tags)
	
	
	def search(self, text=None, category=None, tag=None):
		results = []
		for day in self.days:
			result = None
			if text:
				result = day.search_text(text)
			elif category:
				result = day.search_category(category)
			elif tag:
				result = day.search_tag(tag)
			
			if result:
				if category:
					results.extend(result)
				else:
					results.append(result)
					
		return results
	
	
	def _get_all_edited_days(self):
		# The day being edited counts too
		if self.frame:
			self.save_old_day()
			
		days = []
		for month in self.months.values():
			days_in_month = month.days.values()
			
			# Filter out days without content
			days_in_month = filter(lambda day: not day.empty, days_in_month)
			days.extend(days_in_month)
		return days
	days = property(_get_all_edited_days)
	
	
	def get_word_count_dict(self, type):
		'''
		Returns a dictionary mapping the words to their number of appearance
		'''
		word_dict = collections.defaultdict(int)
		for day in self.days:
			if type == 'word':
				words = day.words
			if type == 'category':
				words = day.node_names
			if type == 'tag':
				words = day.tags
			
			for word in words:
				word_dict[word.lower()] += 1
		return word_dict
	
	def go_to_first_empty_day(self):
		if len(self.sorted_days) == 0:
			return datetime.date.today()
		
		last_edited_day = self.sorted_days[-1]
		first_empty_date = last_edited_day.date + dates.one_day
		self.change_date(first_empty_date)
			
	
	def add_instruction_content(self):
		self.go_to_first_empty_day()
		current_date = self.date
		
		for example_day in info.example_content:
			self.day.content = example_day
			self.frame.set_date(self.month, self.date, self.day)
			self.go_to_next_day()
		
		self.change_date(current_date)
		
	
	
def main():
	start_time = time.time()
	journal = Journal()
	utils.setup_signal_handlers(journal)
	end_time = time.time()
	logging.debug('Start took %s seconds' % (end_time - start_time))
	
	try:
		logging.debug('Trying to enter the gtk main loop')
		gtk.main()
		#logging.debug('Closing logfile')
		#file_logging_stream.close()
	except KeyboardInterrupt:
		# 'Interrupt'
		#journal.save_to_disk()
		sys.exit()
		

if __name__ == '__main__':
	main()
	
